bam_sort.c

```diff
@@ -1611,13 +1633,71 @@ static int change_SO(bam_hdr_t *h, const char *so)
 }
 
 // Function to compare reads and determine which one is < the other
-static inline int bam1_lt(const bam1_p a, const bam1_p b)
+// Handle sort-by-pos and sort-by-name. Used as the secondary sort in bam1_lt_by_tag, if reads are equivalent by tag.
+static inline int bam1_lt_core(const bam1_p a, const bam1_p b)
 {
     if (g_is_by_qname) {
         int t = strnum_cmp(bam_get_qname(a), bam_get_qname(b));
         return (t < 0 || (t == 0 && (a->core.flag&0xc0) < (b->core.flag&0xc0)));
-    } else return (((uint64_t)a->core.tid<<32|(a->core.pos+1)<<1|bam_is_rev(a)) < ((uint64_t)b->core.tid<<32|(b->core.pos+1)<<1|bam_is_rev(b)));
+    } else {
+        return (((uint64_t)a->core.tid<<32|(a->core.pos+1)<<1|bam_is_rev(a)) < ((uint64_t)b->core.tid<<32|(b->core.pos+1)<<1|bam_is_rev(b)));
+    }
+}
+
+// Sort record by tag, using pos or read name as a secondary key if tags are identical. Reads not carrying the tag sort first. 
+// Tags are first sorted by the type character (in case the types differ), or by the appropriate comparator for that type if they agree.
+static inline int bam1_lt_by_tag(const bam1_p a, const bam1_p b) 
+{
+    uint8_t* aux_a = bam_aux_get(a, g_sort_tag);
+    uint8_t* aux_b = bam_aux_get(b, g_sort_tag);
```

Comment:
bam_aux_get() does a linear scan through the read aux data, so putting two of them here is likely to slow the sort down a lot.  It would be better to replace the definition of bam1_p with something like this:
```
typedef struct bam1_p {
    const bam1_t *b;
    const uint8_t *tag;  // Or maybe a ptrdiff_t giving the offset of the tag in the aux data
} bam1_p;
```

The tag pointers can be populated when the bam records are read in, prior to the sort.  This will give a 2*log(n) reduction in the number of calls to bam_aux_get().

You would need to do the same thing during the merge stage, but note that merge can change the aux data in `bam_translate()`.  You would need to set the tag pointer **after** this has happened.